streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -1064,14 +1071,16 @@ private void addAllKeys(final Set<Long> allKeys, final List<KeyValue<Long, Long>
 
     // must be public to allow KafkaProducer to instantiate it
     public static class KeyPartitioner implements Partitioner {
+        private final static LongDeserializer LONG_DESERIALIZER = new LongDeserializer();
+
         @Override
         public int partition(final String topic,
                              final Object key,
                              final byte[] keyBytes,
                              final Object value,
                              final byte[] valueBytes,
                              final Cluster cluster) {
-            return ((Long) key).intValue() % NUM_TOPIC_PARTITIONS;
+            return LONG_DESERIALIZER.deserialize(topic, keyBytes).intValue() % NUM_TOPIC_PARTITIONS;
```

Comment:
This is just needed to make the partitioner work for writing into input topics and to use within KS to write into output topic.

Reply:
Hmm, this sounds to me that the StreamProducer's own `partitionsFor` did not return the num.partitions so we ended up calling `send` with `partition == null`, since otherwise we will get the `partition` as

```
partition = partitioner.partition(topic, key, value, partitions.size());
```

where `partitioner` is the `StreamsPartitioner` and the producer's own partitioner should not be used.

Reply:
I don't think so. The original impl (just for the upstream producer to write into the input topics) was:
```
return ((Long) key).intValue() % NUM_TOPIC_PARTITIONS;
```

However, this assumes that `key` is of type `Long` what is not true when used within streams, because Streams does serialize all data upfront and `key` and `value` type is `byte[]` -- thus, we need to deserialize  to get the original key object.

Reply:
What I was asking is for the necessity of 

```
properties.put(StreamsConfig.producerPrefix(ProducerConfig.PARTITIONER_CLASS_CONFIG), KeyPartitioner.class);
```

As I mentioned, Streams has its own StreamsPartitioner, and if it can get the actual not-null `partition` value passing to the `send` call, then the embedded producer's partitioner would not be used. Maybe I missed something critical here --- did you mean this config is only used for sending data to the source topics? If yes why put it into a streams props?